sphinx_needs/lsp/needs_store.py

```diff
@@ -0,0 +1,111 @@
+# --------------------------------------------------------------------------
+# Licensed under the MIT license.
+# See License.txt in the project root for further license information.
+# --------------------------------------------------------------------------
+
+import importlib.util
+import json
+import logging
+import os
+import sys
+from typing import Any, Dict, List, Optional
+
+from sphinx_needs.lsp.exceptions import NeedlsConfigException
+
+
+class NeedsStore:
+    """Abstraction of needs database."""
+
+    def __init__(self) -> None:
+        self.docs_per_type: Dict[str, List[str]] = {}  # key: need type, val: list of doc names (str)
+        self.needs_per_doc: Dict[Optional[str], List[Dict[Optional[str], Any]]] = {}  # key: docname; val: list of needs
+        self.types: List[str] = []  # list of need types actually defined in needs.json
+        self.declared_types: Dict[str, str] = {}  # types declared in conf.py: {'need directive': 'need title'}
+        self.needs: Dict[Optional[str], Dict[Optional[str], Any]] = {}
+        self.needs_initialized: bool = False
+        self.docs_root: Optional[str] = None
+        self.conf_py_path: str = ""
+
+    def is_setup(self) -> bool:
+        """Return True if database is ready for use."""
+
+        if not self.needs_initialized:
+            return False
+
+        return True
+
+    def set_docs_root(self, docs_root: str) -> None:
+        if not os.path.exists(docs_root):
+            raise FileNotFoundError(f"Docs root directory not found: {docs_root}")
+        self.docs_root = docs_root
+
+    def set_conf_py(self, conf_py_path: str) -> None:
+        if not os.path.exists(conf_py_path):
+            raise FileNotFoundError(f"Given custom configuration file {conf_py_path} not found.")
+        self.conf_py_path = conf_py_path
+
+    def set_declared_types(self) -> None:
+        module_name = "conf"
+        work_dir = os.getcwd()
+        conf_py_path = self.conf_py_path
+        conf_py_dir = os.path.dirname(conf_py_path)
+        conf_py_name = os.path.basename(conf_py_path)
+        os.chdir(conf_py_dir)
+
+        logging.info(f"Loading need_types from {conf_py_name}...")
+
+        spec = importlib.util.spec_from_file_location(module_name, conf_py_path)
+        if spec is None:
+            raise ImportError(f"Created module spec {spec} from {conf_py_name} not exists.")
+
+        module = importlib.util.module_from_spec(spec)
+        sys.modules[module_name] = module
+        try:
+            spec.loader.exec_module(module)  # type: ignore[union-attr]
```

Comment:
ideally this 'type: ignore' comment should be removed as it is obscuring a real potential bug.

`spec.loader` can be `None`, but this case is not explicitly handled in your code- hence the error.